api/src/main/java/net/kyori/adventure/text/renderer/ComponentFlattenerImpl.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * This file is part of adventure, licensed under the MIT License.
+ *
+ * Copyright (c) 2017-2021 KyoriPowered
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.renderer;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.KeybindComponent;
+import net.kyori.adventure.text.ScoreComponent;
+import net.kyori.adventure.text.SelectorComponent;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.TranslatableComponent;
+import net.kyori.adventure.text.format.Style;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+final class ComponentFlattenerImpl implements ComponentFlattener {
+  static final ComponentFlattener BASIC = new Builder()
+    .type(KeybindComponent.class, KeybindComponent::keybind) // IntelliJ is wrong here, this is fine
+    .type(ScoreComponent.class, ScoreComponent::value) // Removed in Vanilla 1.16, but we keep it for backwards compat
+    .type(SelectorComponent.class, SelectorComponent::pattern)
+    .type(TextComponent.class, TextComponent::content)
+    .type(TranslatableComponent.class, TranslatableComponent::key)
+    // The Vanilla game will not print NBT components, expecting those to be resolved with sender context
```

Comment:
From `n.m.n.c.KeybindComponent`:

```java
    public <T> Optional<T> visitSelf(ContentConsumer<T> var1) {
        return this.getNestedComponent().visit(var1);
    }
```

so yes, Vanilla will print out a keybind component's key.

You are correct on the goal here. I'm not ready to add the global translator to this just yet, since some platforms don't fully implement it. We'd also need #259 to implement it appropriately, so that can come later.

Reply:
Then if we're going the vanilla route selector should not display anything?